engine/src/main/java/org/terasology/physics/engine/VoxelWorldSystem.java

```diff
@@ -0,0 +1,145 @@
+
+    @Override
+    public void initialise() {
+        if (physics instanceof BulletPhysics) {
+            btDiscreteDynamicsWorld discreteDynamicsWorld = ((BulletPhysics) physics).getDiscreteDynamicsWorld();
+
+            wrapper = new VoxelCollisionAlgorithmWrapper(ChunkConstants.SIZE_X, ChunkConstants.SIZE_Y, ChunkConstants.SIZE_Z);
+            worldShape = new btVoxelShape(wrapper, new Vector3f(-AABB_SIZE, -AABB_SIZE, -AABB_SIZE), new Vector3f(AABB_SIZE, AABB_SIZE, AABB_SIZE));
+
+            Matrix4f matrix4f = new Matrix4f();
+            btDefaultMotionState blockMotionState = new btDefaultMotionState(matrix4f);
+
+            blockConsInf = new btRigidBodyConstructionInfo(0, blockMotionState, worldShape, new Vector3f());
+            rigidBody = new btRigidBody(blockConsInf);
+            rigidBody.setCollisionFlags(btCollisionObject.CollisionFlags.CF_STATIC_OBJECT | rigidBody.getCollisionFlags());
+            short mask = (short) (~(StandardCollisionGroup.STATIC.getFlag() | StandardCollisionGroup.LIQUID.getFlag()));
+            discreteDynamicsWorld.addRigidBody(rigidBody, physics.combineGroups(StandardCollisionGroup.WORLD), mask);
+        }
+
+        super.initialise();
+    }
+
+    private void tryRegister(Block block) {
+        short id = block.getId();
+        if (!registred.contains(id)) {
+            btCollisionShape shape = ((BulletCollisionShape) block.getCollisionShape()).underlyingShape;
+            btVoxelInfo info = new btVoxelInfo(shape != null && block.isTargetable(), shape != null && !block.isPenetrable(), id, shape, JomlUtil.from(block.getCollisionOffset()), block.getFriction(), block.getRestitution(), block.getFriction());
+            wrapper.setVoxelInfo(info);
+            registred.add(id);
+        }
+    }
+
+
+    @ReceiveEvent(components = {BlockComponent.class})
+    public void onBlockChange(OnChangedBlock event, EntityRef entity) {
+        tryRegister(event.getNewType());
+        wrapper.setBlock(event.getBlockPosition().x, event.getBlockPosition().y, event.getBlockPosition().z, event.getNewType().getId());
+    }
+
+    @ReceiveEvent(components = WorldComponent.class)
+    public void onChunkUloaded(BeforeChunkUnload beforeChunkUnload, EntityRef worldEntity) {
+        Vector3i chunkPos = beforeChunkUnload.getChunkPos();
+        wrapper.freeRegion(chunkPos.x, chunkPos.y, chunkPos.z);
+    }
+
+    @ReceiveEvent(components = {WorldComponent.class})
+    public void onNewChunk(OnChunkLoaded chunkAvailable, EntityRef worldEntity) {
+        Vector3i chunkPos = chunkAvailable.getChunkPos();
+        Chunk chunk = chunkProvider.getChunk(chunkPos);
+        ByteBuffer buffer = ByteBuffer.allocateDirect(2 * (ChunkConstants.SIZE_X * ChunkConstants.SIZE_Y * ChunkConstants.SIZE_Z));
+        buffer.order(ByteOrder.nativeOrder());
+        short[] entries = new short[ChunkConstants.SIZE_X * ChunkConstants.SIZE_Y * ChunkConstants.SIZE_Z];
+        for (int x = 0; x < ChunkConstants.SIZE_X; ++x) {
```

Comment:
Really not sure what this does, but it looks expensive. And is it intentional to use the prefix version of `++` ? That sticks out as unusual and therefore bug prone? Especially if not explained well.